lib/rmg/CRmgTemplateZone.cpp

```diff
@@ -1259,30 +1259,17 @@ void CRmgTemplateZone::randomizeTownType(bool matchUndergroundType)
 	auto townTypesAllowed = (townTypes.size() ? townTypes : getDefaultTownTypes());
 	if(matchUndergroundType && gen->getMapGenOptions().getHasTwoLevels())
 	{
-		if(isUnderground())
-		{
-			//intentionally use AND condition in order to have some choice. If poor random options remains - use pure random
-			//TODO: support new towns from mods to be defined as underground towns
-			if(townTypesAllowed.find(ETownType::DUNGEON) != townTypesAllowed.end() && townTypesAllowed.find(ETownType::INFERNO) != townTypesAllowed.end() && townTypesAllowed.find(ETownType::NECROPOLIS) != townTypesAllowed.end())
-			{
-				townTypesAllowed.erase(ETownType::CASTLE);
-				townTypesAllowed.erase(ETownType::RAMPART);
-				townTypesAllowed.erase(ETownType::TOWER);
-				townTypesAllowed.erase(ETownType::STRONGHOLD);
-				townTypesAllowed.erase(ETownType::FORTRESS);
-				townTypesAllowed.erase(ETownType::CONFLUX);
-			}
-		}
-		else
+		std::set<TFaction> townTypesVerify;
+		for(TFaction factionIdx : townTypesAllowed)
 		{
-			//prohibit on surface if there are options for randomize
-			if(townTypesAllowed.size()>=6)
+			bool preferUnderground = (*VLC->townh)[factionIdx]->preferUndergroundPlacement;
```

Comment:
Will VLC->factions()->getById(factionIdx) work?

Reply:
Yes, I do not know if it will even work. This refactoring with services is quite new. Was merged about a year ago.